### PR TITLE
Ast tests and fns

### PR DESCRIPTION
**Pull request by jbreckman**
2.0.1

Add date support, pretty printing, and some fixed up parsing

### CODE REVIEW

File: ast.js [hunk 1]
```diff
@@ -1,52 +1,109 @@
...
+  else if (node.equation) {
+    result = `${astToString(node.equation.left, pretty, depth)}${node.operation}${astToString(node.equation.right, pretty, depth)}`;
+  }
+  else if (node.paren) {
+    result = `(${astToString(node.paren, pretty, depth, true)})`;
   }
-  if (node.equation) {
-    return `${astToString(node.equation.left)}${node.operation}${astToString(node.equation.right)}`;
+  else if (node.remoteLink) {
+    result = `\${${astToString(node.remoteLink, pretty, depth+1, true)}}`;
```
Comment (ZJONSSON) [inline ast.js:39]: should this be depth+1?

File: ast.js [hunk 2]
```diff
@@ -1,52 +1,109 @@
...
+const clues = require('clues');
+const MAX_DEPTH = 50;
+const INDENTATION_AT_DEPTH = [...new Array(MAX_DEPTH)].map((d,i) => '  '.repeat(i));
+
+function pathParser(path) {
+  try {
+    return parse(path.trim());
+  }
+  catch (e) {
+    throw e.message; // make sure it isn't `Error`
```
Comment (ZJONSSON) [inline ast.js:11]: :+1:

File: paths.pegjs [hunk 3]
```diff
@@ -45,33 +45,54 @@ ImpliedParenExpr = head:WordOrParen tail:("ᐉ" WordOrParen)+ {
...
-MathExpression = operation:("add"/"sub"/"mul"/"div") "(" path:EquationPartList ")" {
+MathExpression = operation:("add"/"sub"/"mul"/"div") _ "(" _ path:EquationPartList _ ")" {
   return {
     operation,
     math: path
   }
 }
 
-CqExpression = operation:("cq") "(" path:Expression ")" {
+LiteralDate = "date" _ "(" path:StringLiteral ")" {
```
Comment (ZJONSSON) [inline paths.pegjs:61]: might be worth allowing optional formatting as second argument (which is the optional second parameter in the `moment` ) https://momentjs.com/docs/#/parsing/string-format/

File: conditional.js [hunk 4]
```diff
@@ -11,6 +13,20 @@ function inverseIdentity(d) { return !d; }
 function identity(d) { return d; }
 function noop() {}
 
+const MOMENTS_UNITS = {
+  'addmonths': 'months',
+  'adddays': 'days',
+  'addyears': 'years'
```
Comment (ZJONSSON) [inline conditional.js:19]: addweeks?

File: paths.pegjs (hunk 3 above)
Comment (jbreckman) [reply paths.pegjs:61]: oh, interesting.  that wouldn't be hard to get in.

File: conditional.js [hunk 5]
```diff
@@ -11,6 +13,20 @@ function inverseIdentity(d) { return !d; }
...
+};
+const INVALID_DATE = new Date(Number.NaN);
```
Comment (ZJONSSON) [inline conditional.js:21]: :)

File: ast.js (hunk 1 above)
Comment (jbreckman) [reply ast.js:39]: The pretty printing isn't perfect -- and as long as it gets parsed correctly on the other side it doesn't matter THAT much. 

but I think I did this on purpose.

```
property.
  residential.
    transactions.
      all.
        select.
          amount<add(
            5, 
            ${global_app.
                data.
                  APPRAISAL}
          )
```
vs

```
property.
  residential.
    transactions.
      all.
        select.
          amount<add(
            5, 
            ${global_app.
              data.
                APPRAISAL}
          )
```

Comment (jbreckman) [reply ast.js:39]: you can see the `${` makes things a little weird with indentation

File: test/ast-test.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+                0.
+                  buyer.
+                    1.
+                      person.
+                        income.
+                          all`);
+
+  let broken = 'property.residential.something.add(someweird,(fsde.sjdrfkl.fsdjkl)|sub(5,(cq(a.b).solve.add((a.b.d), if((a.b.c=5),"yo",${someNestedThing.that.is.deep}),3,2)))).foop.-31=23=fdsjklfajZXZZzzffgjdsktgsk';
+  try {
+    confirmMatches(broken);
```
Comment (ZJONSSON) [inline test/ast-test.js:47]: We could use t.throws https://node-tap.org/docs/api/asserts/#tthrowsfn-expectederror-message-extra

File: test/date-test.js [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        .then(function(d) {
+          t.ok(Query.isPrototypeOf(d),'result does not have a Query prototype');
+          t.same(d.length,1);
+          t.same(d[0].Aspect,'Infrastructure');
+        });
+    });
+
+    t.test('basic filters - nested Date path',{autoend:true},function(t) {
+      let f = Object.create(data);
+      f.dynamic = { date: "DDMMYY" };
```
Comment (ZJONSSON) [inline test/date-test.js:70]: f.dynamic not needed

File: test/select-test.js [hunk 8]
```diff
@@ -148,8 +148,8 @@ t.test('select',{autoend:true},function(t) {
         .then(function(d) {
           t.same(d.length,data.length);
           d.forEach(function(d,i) {
-            t.same(d['add(Value|3)'],data[i].Value+3);
-            t.same(d['sub(Value|5)'],data[i].Value-5);
+            t.same(d['add(Value,3)'],data[i].Value+3);
```
Comment (ZJONSSON) [inline test/select-test.js:151]: :+1:

**Pull request merged**